crates/matrix-sdk-ui/src/room_list_service/sorters/recency.rs

```diff
@@ -31,43 +31,143 @@ where
+                right @ Some(Score::RecencyStamp(_)) | right @ None,
+            ) => (left.map(Score::into_u64), right.map(Score::into_u64)),
+
+            // Scores of different natures.
+            (_, _) => {
+                unreachable!("`Score::TimeStamp` cannot be compared to a `Score::RecencyStamp`")
+            }
+        }
+    }
+}
+
+/// Represent a score from either a latest event's timestamp or a room's recency
+/// stamp.
+#[derive(Debug)]
+enum Score {
+    /// A latest event's timestamp.
+    Timestamp(u64),
+
+    /// A room's recency stamp.
+    RecencyStamp(u64),
+}
+
+impl Score {
+    fn into_u64(self) -> u64 {
+        match self {
+            Self::Timestamp(u64) => u64,
+            Self::RecencyStamp(u64) => u64,
+        }
+    }
+}
+
+/// Extract the _scores_ from either the [`RoomInfo::new_latest_event`]
+/// or from [`RoomInfo::recency_stamp`].
+fn extract_scores(left: &RoomListItem, right: &RoomListItem) -> Scores {
+    // Warning 1.
+    //
+    // Be careful. This method is called **a lot** in the context of a sorter. Using
+    // `Room::new_latest_event` would be dramatic as it returns a clone of the
+    // `LatestEventValue`. It's better to use the more specific method
+    // `Room::new_latest_event_timestamp`, where the value is cached in
+    // `RoomListItem::cached_latest_event_timestamp`.
+
+    // Warning 2.
+    //
+    // A `RoomListItem` must have a unique score when sorting. Its `Score`
+    // must always be the same while sorting the rooms. Thus, the following
+    // rules must apply:
+    //
+    // - Nominal case: Two rooms with a latest event can be compared together based
+    //   on their latest event's timestamp,
+    // - Case #1: If a room has a latest event, but the other doesn't have one, the
+    //   first room has a score but the other doesn't have one,
+    // - Case #2: If none of the room has a latest event, we fallback to the recency
+    //   stamp for both rooms.
+    //
+    // The most important aspect is: if room returns its latest event's
+    // timestamp or its recency stamp, _once_, it must return it every time
+    // it's compared to another room, if possible, whatever the room is,
+    // otherwise it must return `None`.
+    match (left.cached_latest_event_timestamp, right.cached_latest_event_timestamp) {
+        // Nominal case.
+        (left @ Some(_), right @ Some(_)) |
+        // Case #1.
+        (left @ None, right @ Some(_)) |
+        (left @ Some(_), right @ None)
+        => {
+            Scores(
+                left.map(|ts| ts.get().into()).map(Score::Timestamp),
+                right.map(|ts| ts.get().into()).map(Score::Timestamp)
+            )
+        }
+
+        // Case #2.
+        (None, None) => {
+            Scores(
+                left.cached_recency_stamp.map(Into::into).map(Score::RecencyStamp),
+                right.cached_recency_stamp.map(Into::into).map(Score::RecencyStamp)
+            )
+        }
```

Comment:
Not sure a match is the best in terms of readability here? sounds like

```rust
if left.cached_latest_event_timestamp.is_some() || right.cached_latest_event_timestamp.is_some() {
 // use timestamps
} else {
 // use recency stamps
}
```

is shorter and easier on my smol brain. What do you think?